src/Magento2Connection.php

```diff
@@ -91,21 +56,57 @@ public function getPort(): string
 
     public function getDatabase(): string
     {
-        return $this->getDetails()['dbname'];
+        return $this->getDetails()['dbname'] ?? '';
     }
 
     public function getUser(): string
     {
-        return $this->getDetails()['username'];
+        return $this->getDetails()['username'] ?? '';
     }
 
     public function getPassword(): string
     {
-        return $this->getDetails()['password'];
+        return $this->getDetails()['password'] ?? '';
     }
 
+    /**
+     * @return array<string, array<string, string[]>>
+     */
     public function getPreserve(): array
     {
         return $this->settings['preserve'] ?? [];
     }
+
+    /**
+     * @return array<string, string>
+     */
+    private function getDetails(): array
+    {
+        if (is_array($this->connectionDetails)) {
+            return $this->connectionDetails;
+        }
+
+        if (!isset($this->settings['path'])) {
+            $this->connectionDetails = [];
```

Comment:
@michalbiarda Is there any specific reason why we didn't use inline return here in this if statement and other ones below?

Reply:
Yes. In case of any issue we are setting empty array to `$this->connectionDetails`. At the beginning of this method we are checking if the value is array (be default it's NULL).